### PR TITLE
feat: Add support for schema subject naming strategies

### PR DESCRIPTION
**Pull request by Athosone**
## Summary

This pull request implements support for configurable schema subject naming strategies in NS4Kafka, allowing namespaces to enforce different schema naming conventions beyond the default TopicNameStrategy.

### Core Implementation
- **New enum `SubjectNameStrategy`**: Defines the three main Confluent schema naming strategies:
  - `TOPIC_NAME` (default): {topic}-key/{topic}-value format
  - `TOPIC_RECORD_NAME`: {topic}-{recordName} format  
  - `RECORD_NAME`: {recordName} format

- **Enhanced validation framework**: Added `SchemaSubjectNameValidator` with comprehensive validation logic for all naming strategies, including:
  - AVRO schema record name extraction
  - Pattern matching for each strategy type
  - Support for qualified record names

### Integration Points
- **Topic configuration**: Added `value.subject.name.strategy` configuration support
- **Namespace validation**: Extended namespace specs to define valid naming strategies per namespace
- **Schema service**: Updated schema validation to use configured naming strategies instead of hardcoded TopicNameStrategy
- **Error handling**: Enhanced error messages to show expected formats for configured strategies

### Files Modified
- `SubjectNameStrategy.java` (new): Core enum with strategy definitions and format helpers
- `SchemaSubjectNameValidator.java` (new): Validation logic for all naming strategies  
- `TopicValidator.java`: Added naming strategy configuration validation
- `SchemaService.java`: Integrated new validation logic
- `FormatErrorUtils.java`: Enhanced error messages for naming strategy violations
- `Topic.java`: Added subject name strategy configuration support
- Comprehensive test coverage in `SchemaSubjectNameValidatorTest.java`

## Benefits
- **Flexibility**: Supports all Confluent schema naming strategies
- **Backward compatibility**: Default behavior unchanged (TopicNameStrategy)
- **Namespace-level control**: Administrators can configure allowed strategies per namespace
- **Better validation**: Clear error messages showing expected formats
- **Future-ready**: Extensible architecture for additional naming strategies

## Testing
- Added comprehensive unit tests covering all naming strategies
- Validated AVRO record name extraction with various schema formats
- Integration tests ensure backward compatibility
- Edge cases handled (empty schemas, malformed content, etc.)

This implementation provides the foundation for flexible schema management while maintaining full backward compatibility with existing deployments.

## Co-authors
Co-authored-by: monitorpattern <zoedrey.pro@gmail.com>

### CODE REVIEW

File: src/main/java/com/michelin/ns4kafka/service/SchemaService.java [hunk 1]
```diff
@@ -427,4 +437,11 @@ public Mono<Boolean> existInOldVersions(Namespace namespace, Schema schema, List
                                         }))
                                 .any(subjectComparison -> subjectComparison));
     }
+
+    private List<SubjectNameStrategy> getValidSubjectNameStrategies(Namespace namespace) {
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/service/SchemaService.java]: Add comment on this function

File: src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java [hunk 2]
```diff
@@ -67,7 +67,12 @@ public static TopicValidator makeDefault() {
                         "retention.bytes",
                         ResourceValidator.Range.optionalBetween(-1, 104857600),
                         "preallocate",
-                        ResourceValidator.ValidString.optionalIn("true", "false")))
+                        ResourceValidator.ValidString.optionalIn("true", "false"),
+                        VALUE_SUBJECT_NAME_STRATEGY,
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java]: I know this function is already called from tests, but "just in case" could we do not set a subject name strategy here

File: src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java [hunk 3]
```diff
@@ -96,13 +101,25 @@ public static TopicValidator makeDefaultOneBroker() {
                 .build();
     }
 
+    public List<SubjectNameStrategy> getValidSubjectNameStrategies() {
+        ResourceValidator.Validator namingStrategies =
+                getValidationConstraints().get(VALUE_SUBJECT_NAME_STRATEGY);
+        if (namingStrategies instanceof ResourceValidator.ValidString validString
+                && validString.getValidStrings() != null) {
+            return new ArrayList<>(validString.getValidStrings().stream()
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java]: `new ArrayList` should not be necessary here.

File: src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java [hunk 4]
```diff
@@ -96,13 +101,25 @@ public static TopicValidator makeDefaultOneBroker() {
...
+                    .map(SubjectNameStrategy::fromConfigValue)
+                    .toList());
+        }
+        return new ArrayList<>(List.of(SubjectNameStrategy.DEFAULT));
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java]: Same here `new ArrayList` should not be necessary

File: src/main/java/com/michelin/ns4kafka/service/SchemaService.java [hunk 5]
```diff
@@ -196,12 +200,16 @@ public Mono<Schema> getSubjectLatestVersion(Namespace namespace, String subject)
     public Mono<List<String>> validateSchema(Namespace namespace, Schema schema) {
         return Mono.defer(() -> {
             List<String> validationErrors = new ArrayList<>();
+            List<SubjectNameStrategy> namingStrategies = getValidSubjectNameStrategies(namespace);
+            String subjectName = schema.getMetadata().getName();
+            boolean isValid = SchemaSubjectNameValidator.validateSubjectName(
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/service/SchemaService.java]: Just pass the whole `schema` object here instead of these 3 parameters `schema.getMetadata().getName()`, `schema.getSpec().getSchema()`, `schema.getSpec().getSchemaType()`. It'll make it clearer.

Also, we use to define validation functions in services, so `validateSubjectName` can just be a function of the `SchemaService`.

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+     */
+    public static boolean validateSubjectName(
+            String subjectName,
+            List<SubjectNameStrategy> validStrategies,
+            String schemaContent,
+            Schema.SchemaType schemaType) {
+        if (subjectName == null || subjectName.trim().isEmpty()) {
+            return false;
+        }
+        for (SubjectNameStrategy strategy : validStrategies) {
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: I think this can be done in one line:

```java
return validStrategies.stream().anyMatch(strategy ->
                validateSubjectNameWithStrategy(subjectName, strategy, schemaContent, schemaType));
```

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            if (validateSubjectNameWithStrategy(subjectName, strategy, schemaContent, schemaType)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static boolean validateSubjectNameWithStrategy(
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: Add comment to this function.

The link below can be move in a `@see` tag as well like

```java
@see <a href="https://github.com/confluentinc/schema-registry/blob/master/schema-serializer/src/main/java/io/confluent/kafka/serializers/subject">Schema Registry strategies</a>
```

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            String subjectName, SubjectNameStrategy strategy, String schemaContent, Schema.SchemaType schemaType) {
+        // https://github.com/confluentinc/schema-registry/blob/master/schema-serializer/src/main/java/io/confluent/kafka/serializers/subject
+        switch (strategy) {
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: I think this `switch-case` could be written using the `return-switch` pattern:

```java
return switch (strategy) {
    case TOPIC_NAME -> {
        String topicName = extractTopicName(subjectName, strategy).orElse("");
        yield subjectName.equals(topicName + "-key") || subjectName.equals(topicName + "-value");
    }
    case TOPIC_RECORD_NAME -> {
        String topicName = extractTopicName(subjectName, strategy).orElse("");
        Optional<String> recordName = extractRecordName(schemaContent, schemaType);
        yield recordName.isPresent() && subjectName.equals(topicName + "-" + recordName.get());
    }
    case RECORD_NAME -> {
        Optional<String> recordName = extractRecordName(schemaContent, schemaType);
        yield recordName.isPresent() && subjectName.equals(recordName.get());
    }
};
```

Also, feel free to pass the whole `schema` in parameter to avoid having so many parameters.

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+            case TOPIC_NAME:
+                String topicName = extractTopicName(subjectName, strategy).orElse("");
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: ⚠️ Here, we’re just extracting the topic name from the subject and verifying that the subject name equals the extracted topic name plus '-key' or '-value', right?

I think it does not make really sense to extract the topic name for this verification we should just verify that the subject name ends with "-key" or "-value" like it was done before.

Same for TOPIC_RECORD_NAME, let's just verify it ends with `-recordName`

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+     * @param schemaType The type of schema (AVRO, JSON, PROTOBUF)
+     * @return Optional containing the record name if found
+     */
+    public static Optional<String> extractRecordName(String schemaContent, Schema.SchemaType schemaType) {
+        if (schemaContent == null || schemaContent.trim().isEmpty()) {
+            return Optional.empty();
+        }
+
+        try {
+            switch (schemaType) {
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: The IDE is yelling that a single condition `switch-case` should be replaced with an `if`, so let's make it happy.

⚠️ I think the whole `extractAvroRecordName` function can be replaced with `new AvroSchema(schemaContent).name()` which return `namespace + name` for us 😅.

So overall, extracting the record name from the schema can just be:

```java
if (schemaContent == null || schemaContent.trim().isEmpty()) {
    return Optional.empty();
}

if (schemaType == Schema.SchemaType.AVRO) {
    return Optional.of(new AvroSchema(schemaContent).name());
}

return Optional.empty();
```

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+import java.util.List;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+
+/** Validator for schema subject names based on different naming strategies. */
+@Slf4j
+public final class SchemaSubjectNameValidator {
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    private SchemaSubjectNameValidator() {}
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: We use the `@NoArgsConstructor(access = AccessLevel.PRIVATE)` convention in this project instead

File: src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java [hunk 12]
```diff
@@ -96,13 +101,25 @@ public static TopicValidator makeDefaultOneBroker() {
                 .build();
     }
 
+    public List<SubjectNameStrategy> getValidSubjectNameStrategies() {
+        ResourceValidator.Validator namingStrategies =
+                getValidationConstraints().get(VALUE_SUBJECT_NAME_STRATEGY);
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java]: ⚠️ Be careful here:
- If the given subject name ends with `-key` we should verify it against the `KEY_SUBJECT_NAME_STRATEGY`.
- If the given subject name is `RecordName` or `TopicRecordName`, we should verify it against both `KEY_SUBJECT_NAME_STRATEGY` or `VALUE_SUBJECT_NAME_STRATEGY`.

Currently, if a namespace has the following topic validation rules:
- `VALUE_SUBJECT_NAME_STRATEGY` authorized for `TopicName`
- `KEY_SUBJECT_NAME_STRATEGY` authorized for `RecordName` 
And a schema following the `RecordName` strategy is being deployed, the user will be denied even though they're authorized to deploy that schema, right?

File: src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    TOPIC_NAME("io.confluent.kafka.serializers.subject.TopicNameStrategy"),
+    TOPIC_RECORD_NAME("io.confluent.kafka.serializers.subject.TopicRecordNameStrategy"),
+    RECORD_NAME("io.confluent.kafka.serializers.subject.RecordNameStrategy");
+
+    private final String STRATEGY_PREFIX = "io.confluent.kafka.serializers.subject";
+    public static final SubjectNameStrategy DEFAULT = SubjectNameStrategy.TOPIC_NAME;
+
+    private final String value;
+
+    SubjectNameStrategy(String value) {
```
Comment (ThomasCAI-mlv) [inline src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java]: Maybe can be replaced by @AllArgsConstructor on the enum class

File: src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+     * @return SubjectNameStrategy enum value from given confluent's strategy value
+     */
+    public static SubjectNameStrategy fromConfigValue(final String strategyRealValue) {
+        return Arrays.stream(values())
+                .filter(s -> s.value.equals(strategyRealValue))
+                .findFirst()
+                .orElseThrow();
+    }
+
+    public String toShortName() {
```
Comment (ThomasCAI-mlv) [inline src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java]: I find `replace` a bit more readable:
`return value.replace(STRATEGY_PREFIX, "");`
What do you think?

 `.` can be added at the end of the prefix:
`STRATEGY_PREFIX = "io.confluent.kafka.serializers.subject."`

File: src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+                return "{recordName}";
+            default:
+                throw new IllegalArgumentException("Unknown SubjectNameStrategy: " + this);
+        }
+    }
+
+    /**
+     * Return SubjectNameStrategy enum value corresponding to confluent's strategy value given as a String
+     *
+     * @param strategyRealValue
```
Comment (ThomasCAI-mlv) [inline src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java]: The description of strategyRealValue is missing

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java (hunk 9 above)
Comment (ThomasCAI-mlv) [reply src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: Agreed, in the `TOPIC_NAME` case, it is equivalent to checking that the subject name ends with "-key" or "-value"
`return subjectName.endsWith("-value") || subjectName.endsWith("-key");`

And the `TOPIC_RECORD_NAME` case is equivalent to checking that the subject name ends with "-" + recordName.get()

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    /**
+     * Extracts the topic name from a subject name based on the naming strategy.
+     *
+     * @param subjectName The subject name (assumed to be not empty)
+     * @param strategy The naming strategy
+     * @return The topic name if it can be determined
+     */
+    public static Optional<String> extractTopicName(String subjectName, SubjectNameStrategy strategy) {
```
Comment (ThomasCAI-mlv) [inline src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: With the feedback on `validateSubjectNameWithStrategy()`, this method would be used only once, you might consider putting its content in `extractTopicName(String subjectName, List<SubjectNameStrategy> strategies)`

File: src/main/java/com/michelin/ns4kafka/service/SchemaService.java [hunk 17]
```diff
@@ -427,4 +437,11 @@ public Mono<Boolean> existInOldVersions(Namespace namespace, Schema schema, List
                                         }))
                                 .any(subjectComparison -> subjectComparison));
     }
+
+    private List<SubjectNameStrategy> getValidSubjectNameStrategies(Namespace namespace) {
+        if (namespace.getSpec().getTopicValidator() != null) {
+            return namespace.getSpec().getTopicValidator().getValidSubjectNameStrategies();
+        }
+        return List.of(SubjectNameStrategy.DEFAULT);
```
Comment (ThomasCAI-mlv) [inline src/main/java/com/michelin/ns4kafka/service/SchemaService.java]: If no validator is defined on the namespace, it would mean all subject name strategies are allowed in the namespace, so I suggest all strategies to be in the list

File: src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java [hunk 18]
```diff
@@ -96,13 +101,25 @@ public static TopicValidator makeDefaultOneBroker() {
                 .build();
     }
 
+    public List<SubjectNameStrategy> getValidSubjectNameStrategies() {
```
Comment (ThomasCAI-mlv) [inline src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java]: Missing the javadoc

File: src/test/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidatorTest.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+import com.michelin.ns4kafka.model.schema.Schema;
+import com.michelin.ns4kafka.model.schema.SubjectNameStrategy;
+import java.util.List;
+import java.util.Optional;
+import org.junit.jupiter.api.Test;
+
+class SchemaSubjectNameValidatorTest {
+
+    @Test
+    void testValidateSubjectName_TopicNameStrategy_Valid() {
```
Comment (ThomasCAI-mlv) [inline src/test/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidatorTest.java]: Can you follow the naming convention `shouldActLikeThisWhenThatCondition()` or `shouldActLikeThisIfThatCondition()` or `shouldActLikeThisForThatConstraint()` for test methods?

I suggest `shouldValidateSubjectForTopicNameStrategy()` and `shouldNotValidateSubjectForTopicNameStrategy()`

File: src/test/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidatorTest.java [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Test
+    void testValidateSubjectName_RecordNameStrategy_Valid() {
+        String subject = "User";
+        String schemaContent = "{\"name\":\"User\"}";
+        boolean result = SchemaSubjectNameValidator.validateSubjectName(
+                subject, List.of(SubjectNameStrategy.RECORD_NAME), schemaContent, Schema.SchemaType.AVRO);
+        assertTrue(result);
+    }
+
```
Comment (ThomasCAI-mlv) [inline src/test/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidatorTest.java]: Might be better to add an "OK" test with multiple strategies and a "KO" test with multiple strategies

File: src/test/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidatorTest.java [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+    @Test
+    void testValidateSubjectName_TopicRecordNameStrategy_Valid() {
+        String subject = "mytopic-namespace.User";
+        String schemaContent = "{\"name\":\"User\", \"namespace\":\"namespace\"}";
+        boolean result = SchemaSubjectNameValidator.validateSubjectName(
+                subject, List.of(SubjectNameStrategy.TOPIC_RECORD_NAME), schemaContent, Schema.SchemaType.AVRO);
+        assertTrue(result);
+    }
```
Comment (ThomasCAI-mlv) [inline src/test/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidatorTest.java]: For TopicRecordName, might be good to test:
- KO when the subject name has no `-`
- OK when the subject name has multiple `-`
In addition to the simple cases:
- OK when the subject name has one `-`
- KO when the subject name does not match the schema namespace / name

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java [hunk 22]
```diff
@@ -0,0 +1,160 @@
+/*
```
Comment (ThomasCAI-mlv) [inline src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: I'm not sure the content of this file should be in a "Validator" class, it makes more sense in the "SchemaService". What do you think?

**Review by ThomasCAI-mlv — COMMENTED** [review]
Here's my review, feel free to challenge :)

File: src/main/java/com/michelin/ns4kafka/service/SchemaService.java (hunk 17 above)
Comment (Athosone) [reply src/main/java/com/michelin/ns4kafka/service/SchemaService.java]: In this case we want to maintain the current behavior which was using the `TopicName` strategy.

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java (hunk 10 above)
Comment (Athosone) [reply src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: Actually it throws an exception when the references aren't resolved which break existing tests and current behavior.
Also it would throw when parsing union of reference :(

Comment (monitorpattern) [reply src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: Hence, we hesitate between setting a constant value for the record name in the case of union references that do not have namespace/name, and letting the namespace/name parsing. In either case, we need to use the json parser. What do you think?

File: src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+import lombok.AllArgsConstructor;
+
+/** Schema subject naming strategies supported by Schema Registry. */
+@AllArgsConstructor
+public enum SubjectNameStrategy {
+    TOPIC_NAME("io.confluent.kafka.serializers.subject.TopicNameStrategy"),
+    TOPIC_RECORD_NAME("io.confluent.kafka.serializers.subject.TopicRecordNameStrategy"),
+    RECORD_NAME("io.confluent.kafka.serializers.subject.RecordNameStrategy");
+
+    private final String STRATEGY_PREFIX = "io.confluent.kafka.serializers.subject.";
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java]: `static` is missing: `private static final`

File: src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public String toString() {
+        return value;
+    }
+
+    /**
+     * @return The format for subject value (i.e. the SchemaResource metadata name) according to subject name strategy
+     */
+    public String toExpectedFormat() {
+        switch (this) {
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java]: You can use `return-switch` here like:

```java
return switch (this) {
    case TOPIC_NAME -> "{topic}-{key|value}";
    case TOPIC_RECORD_NAME -> "{topic}-{recordName}";
    case RECORD_NAME -> "{recordName}";
};
```

File: src/main/java/com/michelin/ns4kafka/validation/ValidSubjectNameStrategies.java [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.stream.Stream;
+import lombok.AllArgsConstructor;
+
+/**
+ * Valid subject name strategies for value and key. This class is used to validate the subject name strategies for Kafka
+ * topics.
+ */
+@AllArgsConstructor
+public class ValidSubjectNameStrategies {
+    public List<SubjectNameStrategy> validValueStrategies;
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/validation/ValidSubjectNameStrategies.java]: Make both attributes `private` and define a `@Getter` on the class

File: src/main/java/com/michelin/ns4kafka/service/SchemaService.java [hunk 26]
```diff
@@ -427,4 +442,151 @@ public Mono<Boolean> existInOldVersions(Namespace namespace, Schema schema, List
...
+
+    /**
+     * Validates that a schema subject name follows the specified naming strategy.
+     *
+     * @param schema The schema for which a subject name needs to be validated
+     * @return true if the subject name is valid for any of the strategies, false otherwise
+     */
+    public static boolean validateSubjectName(ValidSubjectNameStrategies validStrategies, Schema schema) {
+        if (schema.getMetadata().getName().endsWith("-key")) {
+            return validStrategies.validKeyStrategies.stream()
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/service/SchemaService.java]: Use getter here, based on my comment above: `validStrategies.getValidKeyStrategies()`

File: src/main/java/com/michelin/ns4kafka/service/SchemaService.java [hunk 27]
```diff
@@ -427,4 +442,151 @@ public Mono<Boolean> existInOldVersions(Namespace namespace, Schema schema, List
...
+     *     Registry strategies</a>
+     */
+    public static boolean validateSubjectNameWithStrategy(SubjectNameStrategy strategy, Schema schema) {
+        String subjectName = schema.getMetadata().getName();
+        Schema.SchemaType schemaType = schema.getSpec().getSchemaType();
+        String schemaContent = schema.getSpec().getSchema();
+        if (subjectName == null || subjectName.trim().isEmpty()) {
+            return false;
+        }
+        switch (strategy) {
```
Comment (loicgreffier) [inline src/main/java/com/michelin/ns4kafka/service/SchemaService.java]: `return-switch` here like:

```java
return switch (strategy) {
    case TOPIC_NAME -> subjectName.endsWith("-key") || subjectName.endsWith("-value");
    case TOPIC_RECORD_NAME -> {
        Optional<String> recordName = extractRecordName(schemaContent, schemaType);
        yield recordName.isPresent()
            && (recordName.get() == UNION_AVRO_RECORD_CLASS_NAME
            || subjectName.endsWith("-" + recordName.get()));
    }
    case RECORD_NAME -> {
        Optional<String> recordNameOnly = extractRecordName(schemaContent, schemaType);
        yield recordNameOnly.isPresent() && subjectName.equals(recordNameOnly.get());
    }
};
```

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java (hunk 10 above)
Comment (loicgreffier) [reply src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: This should not be a big issue. All functions are available in the `SchemaService` to build a new `AvroSchema` even with some references:

```java
getSchemaReferences(schema, namespace)
.map(schemaRefs -> new AvroSchema(
    schema.getSpec().getSchema(),
    getReferences(schema),
    schemaRefs,
    null
)
.name())
```

- Call the `getSchemaReferences()` function. The schema and namespace are needed
- Build a `new AvroSchema(...)` with the references

As `getSchemaReferences` is asynchronous, `extractRecordName` should now return a `Mono<String>` as well as all calling methods

File: src/test/java/com/michelin/ns4kafka/service/SchemaServiceTest.java [hunk 28]
```diff
@@ -21,10 +21,8 @@
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.anyString;
-import static org.mockito.Mockito.eq;
```
Comment (ThomasCAI-mlv) [inline src/test/java/com/michelin/ns4kafka/service/SchemaServiceTest.java:25]: Keep the non-wildcard import would be better

File: src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java [hunk 29]
```diff
@@ -0,0 +1,80 @@
...
+    public String toString() {
+        return value;
+    }
+
+    /**
+     * Get the default SubjectNameStrategy.
+     *
+     * @return The default SubjectNameStrategy
+     */
+    public static List<SubjectNameStrategy> defaultStrategy() {
```
Comment (ThomasCAI-mlv) [inline src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java]: Would `defaultStrategies` be a better name?

File: src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java [hunk 30]
```diff
@@ -24,8 +24,7 @@
 import static com.michelin.ns4kafka.util.FormatErrorUtils.invalidNameSpecChars;
 import static com.michelin.ns4kafka.util.FormatErrorUtils.invalidTopicName;
 import static com.michelin.ns4kafka.util.FormatErrorUtils.invalidTopicSpec;
-import static com.michelin.ns4kafka.util.config.TopicConfig.PARTITIONS;
-import static com.michelin.ns4kafka.util.config.TopicConfig.REPLICATION_FACTOR;
```
Comment (ThomasCAI-mlv) [inline src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java:28]: Keep non-wildcard import

File: src/test/java/com/michelin/ns4kafka/integration/SchemaSubjectNameStrategyIntegrationTest.java [hunk 31]
```diff
@@ -0,0 +1,237 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.michelin.ns4kafka.integration;
+
+import static com.michelin.ns4kafka.util.config.TopicConfig.*;
```
Comment (ThomasCAI-mlv) [inline src/test/java/com/michelin/ns4kafka/integration/SchemaSubjectNameStrategyIntegrationTest.java]: Non-wildcard import

File: src/main/java/com/michelin/ns4kafka/service/SchemaService.java (hunk 17 above)
Comment (loicgreffier) [reply src/main/java/com/michelin/ns4kafka/service/SchemaService.java]: @ThomasCAI-mlv Agree with @Athosone. Currently, only `TopicNameStrategy` is allowed by default, this should remain the case when nothing is specified at the namespace level

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java (hunk 16 above)
Comment (loicgreffier) [reply src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: Fixed by the recent updates. It's now part of `matchesStrategy` method

File: src/test/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidatorTest.java (hunk 20 above)
Comment (loicgreffier) [reply src/test/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidatorTest.java]: A bunch of tests has been added to: https://github.com/Athosone/ns4kafka/blob/21d9471e81867b00f10b92a0e07f5494d8a1de4f/src/test/java/com/michelin/ns4kafka/service/SchemaServiceTest.java#L736

File: src/test/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidatorTest.java (hunk 21 above)
Comment (loicgreffier) [reply src/test/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidatorTest.java]: Same as above, most cases are already covered in https://github.com/Athosone/ns4kafka/blob/21d9471e81867b00f10b92a0e07f5494d8a1de4f/src/test/java/com/michelin/ns4kafka/service/SchemaServiceTest.java#L736

If you see any others, feel free to let me know

File: src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java (hunk 22 above)
Comment (loicgreffier) [reply src/main/java/com/michelin/ns4kafka/validation/SchemaSubjectNameValidator.java]: Fixed by the recent updates

File: src/test/java/com/michelin/ns4kafka/service/SchemaServiceTest.java (hunk 28 above)
Comment (loicgreffier) [reply src/test/java/com/michelin/ns4kafka/service/SchemaServiceTest.java:25]: Fixed

File: src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java (hunk 30 above)
Comment (loicgreffier) [reply src/main/java/com/michelin/ns4kafka/validation/TopicValidator.java:28]: Fixed

File: src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java (hunk 29 above)
Comment (loicgreffier) [reply src/main/java/com/michelin/ns4kafka/model/schema/SubjectNameStrategy.java]: Agree, fixed.

File: src/test/java/com/michelin/ns4kafka/integration/SchemaSubjectNameStrategyIntegrationTest.java (hunk 31 above)
Comment (loicgreffier) [reply src/test/java/com/michelin/ns4kafka/integration/SchemaSubjectNameStrategyIntegrationTest.java]: Fixed
